compiler/rustc_hir_analysis/src/collect.rs

```diff
@@ -444,6 +448,80 @@ impl<'tcx> HirTyLowerer<'tcx> for ItemCtxt<'tcx> {
+        // are unable to explicitly disambiguate which impl to use other than by specifying the
+        // self type.
+
+        assert!(!self_ty.has_infer());
+        let infcx = self
+            .tcx()
+            .infer_ctxt()
+            .ignoring_regions()
+            // We use the next trait solver as it results in alias equality being deferred
+            // which we can then just ignore. We don't evaluate any goals in this infcx so
+            // the actual solver behaviour doesn't really matter...
+            .with_next_trait_solver(true)
+            .build(TypingMode::non_body_analysis());
+
+        // We don't just call the normal normalization routine here as we can't provide the
+        // correct `ParamEnv` and it seems dubious to invoke arbitrary trait solving under
+        // the wrong `ParamEnv`. Expanding free aliases doesn't need a `ParamEnv` so we do
+        // this just to make resolution a little bit smarter.
+        let self_ty = self.tcx.expand_free_alias_tys(self_ty);
+        debug!("select_inherent_assoc_candidates: self_ty={:?}", self_ty);
+
+        let mut universes = if self_ty.has_escaping_bound_vars() {
+            vec![None; self_ty.outer_exclusive_binder().as_usize()]
+        } else {
+            vec![]
+        };
+
+        let candidates = rustc_trait_selection::traits::with_replaced_escaping_bound_vars(
+            &infcx,
+            &mut universes,
+            self_ty,
+            |self_ty| {
+                candidates
+                    .into_iter()
+                    .filter(|&&InherentAssocCandidate { impl_, .. }| {
+                        infcx.probe(|_| {
+                            debug!("select_inherent_assoc_candidates: impl_={:?}", impl_);
+                            let impl_ty = self
+                                .tcx()
+                                .type_of(impl_)
+                                .instantiate(self.tcx(), infcx.fresh_args_for_item(span, impl_));
+
+                            // See comment on doing this operation for `self_ty`
+                            let impl_ty = self.tcx.expand_free_alias_tys(impl_ty);
+                            debug!("select_inherent_assoc_candidates: impl_ty={:?}", impl_ty);
+
+                            use rustc_trait_selection::infer::DefineOpaqueTypes;
+                            infcx
+                                // Using an empty `ParamEnv` is pretty weird but we don't actually
+                                // do anything with any of the returned obligations so it doesn't
+                                // matter too much. Building the correct `ParamEnv` would result
+                                // in undesirable query cycles.
+                                .at(&ObligationCause::dummy(), ty::ParamEnv::empty())
+                                .eq(DefineOpaqueTypes::No, self_ty, impl_ty)
+                                .is_ok()
```

Comment:
do we have a test case exercising the difference between this and just using `DeepReject` here?

It feels somewhat iffy to use `eq` and require that to 1) not rely on where-bounds and 2) don't do any eager trait solving

Reply:
Ah so `eq` is slightly better around non lifetime binders I think- `for<T> Foo<T>::IAT` will not unify with `impl Foo<u8>` with `eq` but will with `DeepRejectCtxt` that seems perfectly fine to me though.

Reply:
Oh I guess there's also the recursion depth stuff for deep reject ctxt

Reply:
I'll add a test for this :thinking: I quite like using deep reject instead of hacky type equality where we use a wrong paramenv and such, but I also feel like we should probably be able to handle this case,

Reply:
Added a test